File: packages/server/src/services/hacks.service.ts
import { HackInfo } from '@shared';
import { RawHack } from '../api/defillama.hacks.api.service';
import { toSlug, baseSlug } from '../utils/slug.utils';

export type HackMap = Map<string, HackInfo[]>;

export function buildHackMap(hacks: RawHack[]): HackMap {
  const map: HackMap = new Map();

  const add = (key: string, info: HackInfo) => {
    if (!key) return;
    if (!map.has(key)) map.set(key, []);
    map.get(key)!.push(info);
  };

  for (const h of hacks) {
    const info: HackInfo = {
      name: h.name,
      date: h.date,
      amountUsd: h.amount,
      classification: h.classification ?? 'Unknown',
      technique: h.technique ?? 'Unknown',
      source: h.source,
    };

    const nameSlug = toSlug(h.name);
    add(nameSlug, info);
    add(baseSlug(nameSlug), info);

    if (h.parentProtocolId) {
      const parentSlug = h.parentProtocolId.replace(/^parent#/, '');
      add(parentSlug, info);
      add(baseSlug(parentSlug), info);
    }
  }

  return map;
}

/** Return deduplicated hacks matching the given project slug. */
export function matchHacks(projectSlug: string, hackMap: HackMap): HackInfo[] {
  const seen = new Map<string, HackInfo>();

  const collect = (key: string) => {
    const items = hackMap.get(key);
    if (items) {
      for (const h of items) seen.set(`${h.name}|${h.date}`, h);
    }
  };

  collect(projectSlug);
  collect(baseSlug(projectSlug));

  // Also match any hack key that is a prefix of this project slug
  // e.g. "compound-v3" matches hack key "compound"
  for (const key of hackMap.keys()) {
    if (projectSlug === key || projectSlug.startsWith(key + '-')) {
      collect(key);
    }
  }

  return Array.from(seen.values()).sort((a, b) => b.date - a.date);
}

File: src/Credfeto.Defi.Server/Services/HacksService.cs
using System;
using System.Collections.Generic;
using System.Linq;
using Credfeto.Defi.Server.Models;
using Credfeto.Defi.Server.Utils;

namespace Credfeto.Defi.Server.Services;

/// <summary>
///     Builds and queries a slug-keyed map of protocol hacks.
/// </summary>
internal static class HacksService
{
    /// <summary>
    ///     Builds a map from protocol slug to hack list from raw DefiLlama hack data.
    /// </summary>
    public static IReadOnlyDictionary<string, List<HackInfo>> BuildHackMap(IReadOnlyList<RawHack> hacks)
    {
        Dictionary<string, List<HackInfo>> map = new(StringComparer.OrdinalIgnoreCase);

        foreach (RawHack h in hacks)
        {
            HackInfo info = new()
            {
                Name = h.Name,
                Date = h.Date,
                AmountUsd = h.Amount,
                Classification = h.Classification ?? "Unknown",
                Technique = h.Technique ?? "Unknown",
                Source = h.Source,
            };

            string nameSlug = SlugUtils.ToSlug(h.Name);
            AddToMap(map: map, key: nameSlug, info: info);
            AddToMap(map: map, key: SlugUtils.BaseSlug(nameSlug), info: info);

            if (!string.IsNullOrEmpty(h.ParentProtocolId))
            {
                string parentSlug = h.ParentProtocolId.Replace(
                    oldValue: "parent#",
                    newValue: string.Empty,
                    comparisonType: StringComparison.Ordinal
                );
                AddToMap(map: map, key: parentSlug, info: info);
                AddToMap(map: map, key: SlugUtils.BaseSlug(parentSlug), info: info);
            }
        }

        return map;
    }

    /// <summary>
    ///     Returns deduplicated hacks matching the given project slug.
    /// </summary>
    public static IReadOnlyList<HackInfo> MatchHacks(
        string projectSlug,
        IReadOnlyDictionary<string, List<HackInfo>> hackMap
    )
    {
        Dictionary<string, HackInfo> seen = new(StringComparer.Ordinal);

        Collect(key: projectSlug, hackMap: hackMap, seen: seen);
        Collect(key: SlugUtils.BaseSlug(projectSlug), hackMap: hackMap, seen: seen);

        // Also match any hack key that is a prefix of this project slug
        // e.g. "compound-v3" matches hack key "compound"
        foreach (
            KeyValuePair<string, List<HackInfo>> entry in hackMap.Where(e =>
                string.Equals(a: projectSlug, b: e.Key, comparisonType: StringComparison.OrdinalIgnoreCase)
                || projectSlug.StartsWith(value: e.Key + "-", comparisonType: StringComparison.OrdinalIgnoreCase)
            )
        )
        {
            foreach (HackInfo hack in entry.Value)
            {
                string seenKey = $"{hack.Name}|{hack.Date}";
                _ = seen.TryAdd(key: seenKey, value: hack);
            }
        }

        List<HackInfo> result = [.. seen.Values];
        result.Sort((a, b) => b.Date.CompareTo(a.Date));

        return result;
    }

    private static void AddToMap(Dictionary<string, List<HackInfo>> map, string key, HackInfo info)
    {
        if (string.IsNullOrEmpty(key))
        {
            return;
        }

        if (!map.TryGetValue(key: key, out List<HackInfo>? list))
        {
            list = [];
            map[key] = list;
        }

        list.Add(info);
    }

    private static void Collect(
        string key,
        IReadOnlyDictionary<string, List<HackInfo>> hackMap,
        Dictionary<string, HackInfo> seen
    )
    {
        if (!hackMap.TryGetValue(key: key, out List<HackInfo>? items))
        {
            return;
        }

        foreach (HackInfo h in items)
        {
            string seenKey = $"{h.Name}|{h.Date}";
            _ = seen.TryAdd(key: seenKey, value: h);
        }
    }
}
